db-init.sql

```diff
@@ -0,0 +1,166 @@
+	toPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	fromPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	amount MONEY NOT NULL
+);
+
+-- ProductTypes
+	-- all the different kinds of products
+
+CREATE TABLE ProductTypes (
+	productTypeId SERIAL PRIMARY KEY NOT NULL,
+	productName VARCHAR(255) NOT NULL,
+	productUnits VARCHAR(255) NOT NULL
+);
+
+	CREATE UNIQUE INDEX productNames_lower ON ProductTypes(lower(productName));
+
+	INSERT INTO ProductTypes VALUES (0, 'milk', 'litres');
+
+-- ProductTransactions
+
+CREATE TABLE ProductTransactions (
+	productTransactionUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+	datetime TIMESTAMP WITH TIME ZONE NOT NULL,
+	toPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	fromPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	productTypeId SERIAL REFERENCES ProductTypes(productTypeId) NOT NULL,
+	amountOfProduct REAL NOT NULL,
+	costPerUnit MONEY NOT NULL
+);
+
+-- ProductTypeTransactionAttributes
+	-- the unique kinds of attributes a transaction for a type of product has
+	-- example: we track the milk density and milk temperature for all milk type transactions
+
+CREATE TABLE ProductTypeTransactionAttributes (
+	productTypeId SERIAL REFERENCES ProductTypes(productTypeId) NOT NULL,
+	attrId SERIAL PRIMARY KEY NOT NULL,
+	attrName VARCHAR(255) NOT NULL,
+	UNIQUE(productTypeId, attrId) -- every product type, may not have duplications of types of attributes
+);
+
+	CREATE UNIQUE INDEX productAttrName_lower ON ProductTypeTransactionAttributes(lower(attrName));
+
+	-- milk
+	INSERT INTO ProductTypeTransactionAttributes VALUES(0, 0, 'milkDensityGramsPerMillilitre');
+
+-- TransactionAttributes
+	-- the values for the unique attributes for every transaction
+
+CREATE TABLE ProductTransactionAttribute (
+	productTransactionUuid UUID REFERENCES ProductTransactions(productTransactionUuid) NOT NULL,
+	attrId SERIAL REFERENCES ProductTypeTransactionAttributes(attrId) NOT NULL,
+	attrValue VARCHAR(255) NOT NULL,
+	UNIQUE(productTransactionUuid, attrId) -- for every transaction, all of its attributes must only have one value
+);
+
+-- Loans
+
+CREATE TABLE Loans (
+	loanUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+	moneyTransactionUuid UUID REFERENCES MoneyTransactions(moneyTransactionUuid) NOT NULL,
+	dueDate DATE
+);
+
+-- LoanPayments
+
+CREATE TABLE LoanPayments (
+	loanUuid UUID REFERENCES Loans(loanUuid) NOT NULL,
+	moneyTransactionUuid UUID REFERENCES MoneyTransactions(moneyTransactionUuid) NOT NULL,
+	UNIQUE(loanUuid, moneyTransactionUuid)
+);
+
+-- ProductPayments
+
+CREATE TABLE ProductPayments (
+	productTransactionUuid UUID REFERENCES ProductTransactions(productTransactionUuid) NOT NULL,
+	moneyTransactionUuid UUID REFERENCES MoneyTransactions(moneyTransactionUuid) NOT NULL
+);
+
+-- ProductExports
+
+CREATE TABLE ProductExports (
+	productTypeId SERIAL REFERENCES ProductTypes(productTypeId) NOT NULL,
+	amountOfProduct REAL NOT NULL,
```

Comment:
does this need a corresponding unit attribute?

Reply:
Yes thanks, that was missed

Reply:
I wonder if on LoanPayments if there should be a payment type since they collect by changing the milk rate as well as garnishing payouts.

Reply:
For the ProductExports we may need to add quality to this as well JT is asking Alex about this.

Reply:
Actually we don't need a coressponding unit attribute, you can't figure it out by knowing what type of product it is.